File: src/main/java/com/hazza/leetcode/SegmentThree/No112PathSum.java
package com.hazza.leetcode.SegmentThree;

import com.hazza.leetcode.datastructure.TreeNode;

import java.util.ArrayDeque;
import java.util.Deque;

/**
 * Created with IntelliJ IDEA.
 * Description: https://leetcode.com/problems/path-sum/description/
 * User: HazzaCheng
 * Contact: hazzacheng@gmail.com
 * Date: 18-1-24
 * Time: 1:18 PM
 */
public class No112PathSum {

    // solution 1
    public boolean hasPathSum(TreeNode root, int sum) {
        if (root == null) return false;
        sum -= root.val;
        if (root.left == null && root.right == null && sum == 0) return true;
        return hasPathSum(root.left, sum) || hasPathSum(root.right, sum);
    }

    // solution 2
    public boolean hasPathSum1(TreeNode root, int sum) {
        if (root == null) return false;
        Deque<TreeNode> stack = new ArrayDeque<>();
        stack.push(root);
        while (!stack.isEmpty()) {
            TreeNode cur = stack.pop();
            if (cur.left == null && cur.right == null && cur.val == sum)
                return true;
            if (cur.left != null) {
                cur.left.val += cur.val;
                stack.push(cur.left);
            }
            if (cur.right != null) {
                cur.right.val += cur.val;
                stack.push(cur.right);
            }
        }

        return false;
    }
}

File: src/main/python/leetcode_by_python/tree/No112_Path_Sum.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019-10-20
from src.main.python.leetcode_by_python.data_structure.TreeNode import TreeNode


class Solution:
    def hasPathSum1(self, root: TreeNode, sum: int) -> bool:
        if not root:
            return False

        sum -= root.val
        if not root.left and not root.right and sum == 0:
            return True

        return self.hasPathSum1(root.left, sum) or self.hasPathSum1(root.right, sum)

    def hasPathSum2(self, root: TreeNode, sum: int) -> bool:
        if not root:
            return False
        stack = [root]

        while stack:
            cur = stack[-1]
            stack.pop()

            if not cur.left and not cur.right and cur.val == sum:
                return True
            if cur.left:
                cur.left.val += cur.val
                stack.append(cur.left)
            if cur.right:
                cur.right.val += cur.val
                stack.append(cur.right)

        return False
